File: backend/src/main/kotlin/com/couples/recording/domain/Entities.kt
package com.couples.recording.domain

import jakarta.persistence.CollectionTable
import jakarta.persistence.Column
import jakarta.persistence.ElementCollection
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.Table

// The relational schema mirrors the client's single JSON state tree. Rows use
// the same string ids the client used (prefixes p_/ex_/t_/sess_/se_/set_), so
// the reassembled aggregate JSON is byte-for-byte compatible with the old
// localStorage shape. Relationships are kept as plain id strings (no JPA
// associations) to keep the document-style model simple and SQLite-friendly.

@Entity
@Table(name = "person")
class Person(
    @Id var id: String = "",
    var name: String = "",
    var initials: String = "",
    var color: String = "",
    var unit: String = "kg",
    var isOwner: Boolean = false,
    var active: Boolean = true,
)

@Entity
@Table(name = "exercise")
class Exercise(
    @Id var id: String = "",
    var name: String = "",
    var category: String = "",
    var equipment: String = "",
    var tracksWeight: Boolean = true,
    var tracksReps: Boolean = true,
    var tracksDuration: Boolean = false,
    var defaultRestSeconds: Int = 90,
)

@Entity
@Table(name = "person_exercise_profile")
class PersonExerciseProfile(
    @Id var id: String = "",
    var personId: String = "",
    var exerciseId: String = "",
    var restSeconds: Int? = null,
    var machineSetup: String = "",
    var cues: String = "",
)

@Entity
@Table(name = "template")
class Template(
    @Id var id: String = "",
    var name: String = "",
    var defaultMode: String = "alternate",
)

@Entity
@Table(name = "template_exercise")
class TemplateExercise(
    @Id var id: String = "",
    var templateId: String = "",
    var exerciseId: String = "",
    var assignment: String = "both",
    var orderIndex: Int = 0,
    var defaultLoggingMode: String? = null,
)

@Entity
@Table(name = "app_settings")
class AppSettings(
    @Id var id: String = SINGLETON_ID,
    var coupleModeEnabled: Boolean = true,
    var defaultParticipants: String = "both",
    var defaultLoggingStyle: String = "alternate",
    var allowCopyPartnerValues: Boolean = true,
    var showPartnerHistory: Boolean = true,
    var onboarded: Boolean = true,
    var version: Int = 3,
) {
    companion object {
        const val SINGLETON_ID = "app"
    }
}

@Entity
@Table(name = "workout_session")
class WorkoutSession(
    @Id var id: String = "",
    var templateId: String? = null,
    var name: String = "",
    var startTime: Long = 0,
    var endTime: Long? = null,
    var label: String? = null,
    var loggingStyle: String = "alternate",
    var status: String = "active",
    @ElementCollection
    @CollectionTable(name = "session_participant", joinColumns = [JoinColumn(name = "session_id")])
    @Column(name = "person_id")
    var participantIds: MutableList<String> = mutableListOf(),
)

@Entity
@Table(name = "session_exercise")
class SessionExercise(
    @Id var id: String = "",
    var sessionId: String = "",
    var exerciseId: String = "",
    var loggingMode: String = "alternate",
    var activePersonId: String? = null,
    var addedDuringSession: Boolean = false,
    var orderIndex: Int = 0,
)

@Entity
@Table(name = "session_exercise_person")
class SessionExercisePerson(
    @Id var id: String = "",
    var sessionExerciseId: String = "",
    var personId: String = "",
    var status: String = "pending",
    var skipReason: String? = null,
    var substituteExerciseId: String? = null,
    var orderIndex: Int = 0,
)

@Entity
@Table(name = "set_entry")
class SetEntry(
    @Id var id: String = "",
    var sessionId: String = "",
    var sessionExerciseId: String? = null,
    var exerciseId: String = "",
    var personId: String = "",
    var setIndex: Int = 0,
    var weight: Double? = null,
    var reps: Int? = null,
    var duration: Int? = null,
    var setType: String = "working",
    var timestamp: Long? = null,
    var note: String? = null,
)

@Entity
@Table(name = "rest_timer")
class RestTimer(
    @Id var id: String = "",
    var sessionId: String = "",
    var personId: String = "",
    var sessionExerciseId: String = "",
    var startedAt: Long = 0,
    var durationSeconds: Int = 0,
)

File: backend/src/entities.rs
// SeaORM entities, one inline module per table. The schema mirrors the old
// Kotlin/JPA backend (snake_case columns), so the reassembled aggregate JSON is
// identical. Relationships are kept as plain id strings (no SeaORM relations) to
// match the document-style model and keep queries simple.

pub mod person {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "person")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub name: String,
        pub initials: String,
        pub color: String,
        pub unit: String,
        pub is_owner: bool,
        pub active: bool,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod exercise {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "exercise")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub name: String,
        pub category: String,
        pub equipment: String,
        pub tracks_weight: bool,
        pub tracks_reps: bool,
        pub tracks_duration: bool,
        pub default_rest_seconds: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod person_exercise_profile {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "person_exercise_profile")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub person_id: String,
        pub exercise_id: String,
        pub rest_seconds: Option<i32>,
        pub machine_setup: String,
        pub cues: String,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod template {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "template")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub name: String,
        pub default_mode: String,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod template_exercise {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "template_exercise")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub template_id: String,
        pub exercise_id: String,
        pub assignment: String,
        pub order_index: i32,
        pub default_logging_mode: Option<String>,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod app_settings {
    use sea_orm::entity::prelude::*;

    pub const SINGLETON_ID: &str = "app";

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "app_settings")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub couple_mode_enabled: bool,
        pub default_participants: String,
        pub default_logging_style: String,
        pub allow_copy_partner_values: bool,
        pub show_partner_history: bool,
        pub onboarded: bool,
        pub version: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod workout_session {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "workout_session")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub template_id: Option<String>,
        pub name: String,
        pub start_time: i64,
        pub end_time: Option<i64>,
        pub label: Option<String>,
        pub logging_style: String,
        pub status: String,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod session_participant {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "session_participant")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub session_id: String,
        pub person_id: String,
        pub order_index: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod session_exercise {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "session_exercise")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub session_id: String,
        pub exercise_id: String,
        pub logging_mode: String,
        pub variant: String,
        pub active_person_id: Option<String>,
        pub added_during_session: bool,
        pub order_index: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod session_exercise_person {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "session_exercise_person")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub session_exercise_id: String,
        pub person_id: String,
        pub status: String,
        pub skip_reason: Option<String>,
        pub substitute_exercise_id: Option<String>,
        pub order_index: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod set_entry {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "set_entry")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub session_id: String,
        pub session_exercise_id: Option<String>,
        pub exercise_id: String,
        pub person_id: String,
        pub set_index: i32,
        pub weight: Option<f64>,
        pub reps: Option<i32>,
        pub duration: Option<i32>,
        pub set_type: String,
        pub variant: String,
        pub timestamp: Option<i64>,
        pub note: Option<String>,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}

pub mod rest_timer {
    use sea_orm::entity::prelude::*;

    #[derive(Clone, Debug, PartialEq, DeriveEntityModel)]
    #[sea_orm(table_name = "rest_timer")]
    pub struct Model {
        #[sea_orm(primary_key, auto_increment = false)]
        pub id: String,
        pub session_id: String,
        pub person_id: String,
        pub session_exercise_id: String,
        pub started_at: i64,
        pub duration_seconds: i32,
    }
    #[derive(Copy, Clone, Debug, EnumIter, DeriveRelation)]
    pub enum Relation {}
    impl ActiveModelBehavior for ActiveModel {}
}
